hedera-node/hedera-schedule-service-impl/src/main/java/com/hedera/node/app/service/schedule/impl/handlers/ScheduleDeleteHandler.java

```diff
@@ -26,26 +31,65 @@
-     * the handle stage.
-     *
-     * <p>Please note: the method signature is just a placeholder which is most likely going to
-     * change.
+     * <p>Pre-handles a {@link
+     * com.hederahashgraph.api.proto.java.HederaFunctionality#ScheduleDelete} transaction, returning
+     * the metadata required to, at minimum, validate the signatures of all required signing keys.
      *
-     * @param txBody the {@link TransactionBody} with the transaction data
+     * @param txn the {@link TransactionBody} with the transaction data
      * @param payer the {@link AccountID} of the payer
+     * @param keyLookup the {@link AccountKeyLookup} to use for key resolution
      * @return the {@link TransactionMetadata} with all information that needs to be passed to
      *     {@link #handle(TransactionMetadata)}
      * @throws NullPointerException if one of the arguments is {@code null}
      */
     public TransactionMetadata preHandle(
-            @NonNull final TransactionBody txBody, @NonNull final AccountID payer) {
-        throw new UnsupportedOperationException("Not implemented");
+            @NonNull final TransactionBody txn,
+            @NonNull final AccountID payer,
+            @NonNull final AccountKeyLookup keyLookup,
+            @NonNull final ReadableScheduleStore scheduleStore) {
+        final var op = txn.getScheduleDelete();
+        final var id = op.getScheduleID();
+
+        // check for INVALID_SCHEDULE_ID
+        final var scheduleLookupResult = scheduleStore.get(id);
+        if (scheduleLookupResult.isEmpty()) {
+            return new InvalidTransactionMetadata(txn, payer, INVALID_SCHEDULE_ID);
+        }
+
+        // No need to check for SCHEDULE_PENDING_EXPIRATION, SCHEDULE_ALREADY_DELETED,
+        // SCHEDULE_ALREADY_EXECUTED
+        // if any of these are the case then the scheduled tx would not be present in scheduleStore
+
+        // check whether schedule was created with an admin key
+        // if it wasn't, the schedule can't be deleted
+        final var adminKey = scheduleLookupResult.get().adminKey();
+        if (adminKey.isEmpty()) {
+            return new InvalidTransactionMetadata(txn, payer, SCHEDULE_IS_IMMUTABLE);
+        }
+
+        // check whether payer key for this tx matches the admin key for underlying scheduled
+        // transaction
+        final KeyOrLookupFailureReason payerKeyOrLookupFailure = keyLookup.getKey(payer);
+        if (payerKeyOrLookupFailure.failed()) {
+            return new InvalidTransactionMetadata(
+                    txn, payer, payerKeyOrLookupFailure.failureReason());
+        } else {
+            var payerKey = payerKeyOrLookupFailure.key();
+            if (!adminKey.get().equals(payerKey))
+                return new InvalidTransactionMetadata(
+                        txn,
+                        payer,
+                        INVALID_SIGNATURE); // I'm not sure whether this is the correct code
```

Comment:
I think all the logic to be used should be from [here](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/sigs/order/SigRequirements.java#L1222).  This seems different

Reply:
I believe the logic is the same both places. It's a little clearer after the edits that I just checked in.

Reply:
What bothers me is that the specs [here](https://github.com/hashgraph/hedera-services/blob/develop/docs/scheduled-transactions/revised-spec.md#the-scheduledelete-transaction) and [here](https://hashgraph.github.io/hedera-protobufs/#schedule_delete.proto) mention additional return codes that are never actually generated, either in the old or refactored way of doing things, afaict. If a schedule is already deleted or executed, you'll get an `INVALID_SCHEDULE_ID` return code, when you should get  `SCHEDULE_ALREADY_DELETED`, `SCHEDULE_ALREADY_EXECUTED` or `SCHEDULE_PENDING_EXPIRATION`.

Reply:
Per Slack convo with @tinker-michaelj , the specs need to be updated to match current behavior. I will create a separate issue for that.